### ISSUE TITLE
Development Frustration: Microlens lacks mapMOf

### ISSUE THREAD

**Issue by jhance** [yi-editor/yi#879]
Hi, please excuse me while I vent from being unable to write code as nice as I want it to be :(

Is there a way to write monad-forms of `over` with microlens? I keep wanting to use `mapMOf` but after the switch to microlens we can no longer use that function. (Its super useful for yi, because we use lenses everywhere + we use monads everywhere)

Microlens *explicitly* states that they will not include these functions (the ones that end with "Of"), which makes me think there should be a nice readable way to do something like `(mapMOf (each . _1) (readMaybe . T.unpack))` [actually, this itself is bad, but I'm bad with traversals, need to figure that out] for parsing an element of a tuple within a list, but I cannot for the life of me come up with it.

I do agree with points about lens adding lots of dependencies for small functions, though, so I can understand why we want to stick with microlens. Its just really annoying being denied useful functions due to the philosophy of the package.

**Comment by noughtmare** [comment]
They have added `mapAccumLOf` (I don't know a lot about lenses), can you give a source for 
>Microlens explicitly states that they will not include these functions (the ones that end with "Of")

**Comment by noughtmare** [comment]
they have added `mapAccumLOf`, so the statement in the readme may be too general.

Maybe @neongreen can say if he has a nice version of `mapMOf`.

**Comment by jhance** [comment]
Its one of the first things in their design decisions: https://github.com/aelve/microlens

**Comment by neongreen** [comment]
I've already added `mapAccumLOf` so I can add `mapMOf` as well.

**Comment by neongreen** [comment]
On the other hand, I'm wondering. What does `mapMOf` need you that the direct application of the lens doesn't?

```
> both (\x -> [x, x + 1]) (1,3)
[(1,3),(1,4),(2,3),(2,4)]
> mapMOf both (\x -> [x, x + 1]) (1,3)
[(1,3),(1,4),(2,3),(2,4)]
```

**Comment by jhance** [comment]
Readability. I had never actually thought of directly applying the lens instead of using `mapMOf`, if that works then I guess I don't care about the function anymore.

**Comment by neongreen** [comment]
I always apply lenses directly instead of using `mapMOf` (it's a sort of a “standard idiom” with lenses), but okay, I can see how it would be unreadable to some people. Okay, I'll add it (or `traverseOf`, I'm not sure) and update the “design decisions” doc tomorrow.

Regarding the `*Of` functions, the “official statement” on the matter is here: https://github.com/aelve/microlens/issues/79#issuecomment-231720804. I fully agree with “It's just really annoying being denied useful functions due to the philosophy of the package”.

*However*, you should also be aware that there's a limit on how many functions I can add before microlens stops being micro – which is a point in favor of not using microlens for huge applications like yi :)

**Comment by jhance** [comment]
I think traverseOf makes more sense to include

**Comment by neongreen** [comment]
Okay, I've added `traverseOf` and `traverseOf_` (which are the same but for `Applicative`) and released microlens-0.4.6 and microlens-platform-0.3.4.
